libindy/src/services/pool/state_proof/mod.rs

```diff
@@ -477,6 +493,95 @@ fn _verify_proof(proofs_rlp: &[u8], root_hash: &[u8], key: &[u8], expected_value
     }).unwrap_or(false)
 }
 
+fn _verify_proof_range(proofs_rlp: &[u8],
+                       root_hash: &[u8],
+                       prefix: &str,
+                       from: Option<u64>,
+                       next: Option<u64>,
+                       kvs: &Vec<(String, Option<String>)>) -> bool {
+    debug!("verify_proof_range >> from {:?}, prefix {:?}, kvs {:?}", from, prefix, kvs);
+    let nodes: Vec<Node> = UntrustedRlp::new(proofs_rlp).as_list().unwrap_or_default(); //default will cause error below
+    let mut map: TrieDB = HashMap::new();
+    for node in &nodes {
+        map.insert(node.get_hash(), node);
+    }
+    map.get(root_hash).map(|root| {
+        let prev = root.get_node(&map, prefix.as_bytes()).map_err(map_err_err!());
+        let node = if let Ok(Some(node)) = prev {
+            node
+        } else {
+            error!("Prefix not found, aborting");
+            return false
+        };
+        let res = node.get_all_values(&map).map_err(map_err_err!());
+        trace!("All values from trie: {:?}", res);
+        let vals = if let Some(vals) = res.ok() {
+            vals
+        } else {
+            error!("Some errors happened while collecting values from state proof");
+            return false
+        };
+        // Preparation of data for verification
+        // Fetch numerical suffixes
+        let vals_for_sort_check: Vec<Option<(u64, (String, String))>> = vals.into_iter()
+            .filter(|(key, _)| key.starts_with(prefix))
+            .map(|(key, value)| {
+                let no = key.replacen(prefix, "", 1).parse::<u64>();
+                no.ok().map(|a| (a, (key, value)))
+        }).collect();
+        if !vals_for_sort_check.iter().all(|a| a.is_some()) {
+            error!("Some values in state proof are not correlating with state proof rule, aborting.");
+            return false;
+        }
+        let mut vals_for_sort: Vec<(u64, (String, String))> = vals_for_sort_check.into_iter().flat_map(|a| a).collect();
+        // Sort by numerical suffixes in ascending order
+        vals_for_sort.sort_by(|a, b| {
+            let (a_seq_no, _) = a;
+            let (b_seq_no, _) = b;
+            a_seq_no.cmp(b_seq_no)
+        });
+        trace!("Sorted trie values: {:?}", vals_for_sort);
+        // Shift on the left side by from
+        let vals_with_from = if let Some(from_seqno) = from {
+            if let Some((idx, _)) = vals_for_sort.iter().enumerate().find(|(_, (seq_no, _))| *seq_no as u64 >= from_seqno) {
+                vals_for_sort[idx..].to_vec()
+            } else {
+                vec![]
+            }
+        } else {
+            vals_for_sort
+        };
+        // Verification
+        // Check that all values from response match the trie
+        trace!("Got values from trie: {:?}", vals_with_from);
```

Comment:
Can it be simplified to just 2 arrays comparison?

Reply:
simplified to slices comparison